feat: add filter period to iframe src for custom widgets

EdsonNhancale opened the pull request:
Implements [DHIS2-19186](https://dhis2.atlassian.net/browse/DHIS2-19186)
### Key features

1. Support for adding `pe` (period) as a query parameter in the iframe source URL.
2. Dynamically appends the `pe` parameter to the iframe source when the filter is applied.

---

### Description

This update ensures that the `pe` (period) filter is included in the iframe URL for dashboard items when the filter is applied. If the `pe` filter is present in `itemFilters`, the periods are concatenated using `,` as the delimiter and added as a `period` query parameter in the iframe source.


---

### TODO

-   [ ] Verify the behavior with different combinations of filters (FILTER_ORG_UNIT and FILTER_PE).

---

### TESTS

-   [x] Added tests to validate the behavior of the period filter with single and multiple values.
-   [ ] Ensured proper handling when both pe and ou filters are applied together.
-   [x] Included a test to confirm period filter is omitted when empty
-   [ ] Verified the generated iframeSrc matches the expected URL for all scenarios.

---

### Known issues

-   [ ] When using custom widgets installed on the dashboard, the widget cannot currently react to changes in the period (pe) filter. This limitation arises because the pe filter value isn't dynamically communicated to the widget. A follow-up issue may be required to address this behavior.


---

### Screenshots
![403252875-57e89dd7-3737-4297-9a33-5b5ee5181e84](https://github.com/user-attachments/assets/4aad1cf4-2375-4641-9d7a-0e5fe440dc01)

![403267034-3fc7b907-5603-4a61-9131-48e13d97a7a9](https://github.com/user-attachments/assets/0aab0cdd-2184-4db4-8d6d-0643a3bb3a9d)


---

**Example:**
If the `itemFilters` contains:
```javascript
{
    FILTER_ORG_UNIT: [{ id: 'ou1', path: '/root/ou1' }, { id: 'ou2' }],
    FILTER_PE: [{id:'THIS_MONTH'}, {id:'LAST_MONTH'}]
}



[DHIS2-19186]: https://dhis2.atlassian.net/browse/DHIS2-19186?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

EdsonNhancale committed:
feat: implemented searchUrl to be able to access the ou and pe

EdsonNhancale committed:
feat: added pe on iframesrc to use in custom widgets

EdsonNhancale committed:
fix: removed unnecessary imports

EdsonNhancale committed:
fix: removed unnecessary imports

KaiVandivier commented on src/components/Item/AppItem/getIframeSrc.js:
```diff
@@ -11,5 +11,12 @@ export const getIframeSrc = (appDetails, item, itemFilters) => {
 
+    // Add period (pe) to iframeSrc
+    if (itemFilters[FILTER_PE] && itemFilters[FILTER_PE].length) {
+        const peValues = itemFilters[FILTER_PE].map(x => x.id).join(';');
```
To match the format of the org unit parameter, let's use a comma as delimeter:

```suggestion
        const peValues = itemFilters[FILTER_PE].map(x => x.id).join(',');
```

KaiVandivier commented on src/components/Item/AppItem/getIframeSrc.js:
```diff
@@ -11,5 +11,12 @@ export const getIframeSrc = (appDetails, item, itemFilters) => {
+    // Add period (pe) to iframeSrc
+    if (itemFilters[FILTER_PE] && itemFilters[FILTER_PE].length) {
+        const peValues = itemFilters[FILTER_PE].map(x => x.id).join(';');
+        iframeSrc += `&pe=${peValues}`;
```
To match the org unit parameter, let's spell out `period`:
```suggestion
        iframeSrc += `&period=${peValues}`;
```

KaiVandivier commented on src/components/Item/AppItem/getIframeSrc.js:
```diff
@@ -11,5 +11,12 @@ export const getIframeSrc = (appDetails, item, itemFilters) => {
     }
 
+    // Add period (pe) to iframeSrc
+    if (itemFilters[FILTER_PE] && itemFilters[FILTER_PE].length) {
```
I know that this is the same logic as the expression above, but [Sonarqube](https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=3183&id=dhis2_dashboard-app&open=AZRoa7daNcCCOAJdJDaq) recommends using an optional chaining expression here:
```suggestion
    if (itemFilters[FILTER_PE]?.length) {
```

KaiVandivier reviewed (commented):
Thanks for opening this Edson! 🙏  The logic looks good 👍 

There's some things we should do to clean up the PR before we ask analytics folks to review it:
1. Add a more descriptive PR name, using conventional commit style, like "feat: add filter period to iframe src for custom widgets"
2. In the description, some code block styling is wrapping the screenshots, so the images aren't visible -- move those screenshots out of the code block so they're easy to see
3. There are some erroneous changes in `en.pot` and `FilterSelector.js` -- remove those
4. I added some inline comments in the code -- if you use the GitHub interface to apply them, make sure to use commit messages that follow the conventional commit style (start with `chore:`/`fix:`/`refactor:`)

Thanks again for setting this up! Let me know if you have any questions 🙂

EdsonNhancale committed:
refactor: update delimiter for `pe` values in iframe source URL

Changed the delimiter for `pe` values from semicolon (`;`) to comma (`,`), aligning with the standard format for query parameters. This change improves consistency and simplifies downstream processing.

Co-authored-by: Kai Vandivier <49666798+KaiVandivier@users.noreply.github.com>

EdsonNhancale committed:
refactor: use optional chaining to simplify null/undefined checks

Replaced the explicit `if` condition `itemFilters[FILTER_PE] && itemFilters[FILTER_PE].length` with the optional chaining expression `itemFilters[FILTER_PE]?.length`, as recommended by SonarQube.

Co-authored-by: Kai Vandivier <49666798+KaiVandivier@users.noreply.github.com>

EdsonNhancale committed:
refactor: rename `pe` parameter to `period` for clarity and consistency

Updated the query parameter `pe` to `period` in the iframe source URL to improve readability and better match the naming convention of the org unit parameter.

Co-authored-by: Kai Vandivier <49666798+KaiVandivier@users.noreply.github.com>

EdsonNhancale committed:
fix: remove `i18n/en.pot` file to avoid conflicts

Deleted the `i18n/en.pot` file to prevent potential conflicts with other translations or resources. This resolves issues with unnecessary file changes during development and ensures consistency in localization files.

EdsonNhancale committed:
refactor: remove unnecessary changes in `FilterSelector.js`

Cleaned up the `FilterSelector.js` file by removing redundant or unnecessary changes.

EdsonNhancale committed:
fix: fix: remove erroneous changes in `en.pot`

Reverted unintended changes in the `i18n/en.pot`

EdsonNhancale committed:
Merge branch 'iframesrc-add-pe' of https://github.com/EdsonNhancale/dashboard-app into iframesrc-add-pe

EdsonNhancale committed:
fix: remove erroneous changes in `FilterSelector.js`

Reverted unintended changes in the `FilterSelector.js` file

jenniferarnesen commented:
Thank you for submitting this pull request. Could you please have a look at the unit tests in file `__tests__/getIframeSrc.spec.js` and add relevant tests for the new code including:
* only pe filter (including cases with just 1 and >1 filters)
* pe filter and ou filter

EdsonNhancale committed:
test: added unit tests for period (pe) filter functionality in getIframeSrc

- Added tests to validate the behavior of the period (pe) filter with single and multiple values. - Ensured proper handling when both pe and ou filters are applied together. - Included a test to confirm pe filter is omitted when empty. - Verified the generated iframeSrc matches the expected URL for all scenarios.

EdsonNhancale committed:
test: added unit tests for period (pe) filter functionality in getIframeSrc

- Added tests to validate the behavior of the period (pe) filter with single and multiple values. - Ensured proper handling when both pe and ou filters are applied together. - Included a test to confirm pe filter is omitted when empty. - Verified the generated iframeSrc matches the expected URL for all scenarios.

jenniferarnesen commented:
Hi Edson,
There are lint failures. You can ignore the `lint-commit` error (unless you want to rebase and fix it that way) because we'll do a squash commit when the time comes and rewrite the commit message at that point. But the `lint` error has to be fixed.

EdsonNhancale committed:
Merge branch 'iframesrc-add-pe' of https://github.com/EdsonNhancale/dashboard-app into iframesrc-add-pe

edoardo commented:
Hi @EdsonNhancale , before we can merge this there is a conflict with `master` that needs to be resolved.
Could you please merge `master` into your branch and ensure this is the content of `src/components/Item/AppItem/getIframeSrc.js` after the resolution?
Thanks!


```
import { FILTER_ORG_UNIT, FILTER_PE } from '../../../actions/itemFilters.js'
import { getPluginOverrides } from '../../../modules/localStorage.js'

export const getIframeSrc = (item, itemFilters, appDetails = {}) => {
    if (appDetails.appType === 'APP') {
        // pluginOverrides might be in use during development on localhost
        const pluginOverrides = getPluginOverrides()

        if (pluginOverrides?.[appDetails.key]) {
            return pluginOverrides[appDetails.key]
        }

        return appDetails.pluginLaunchUrl
    } else {
        let iframeSrc = `${appDetails.launchUrl}?dashboardItemId=${item.id}`

        if (itemFilters[FILTER_ORG_UNIT]?.length) {
            const ouIds = itemFilters[FILTER_ORG_UNIT].map(({ id, path }) =>
                path ? path.split('/').slice(-1)[0] : id
            )
            iframeSrc += `&userOrgUnit=${ouIds.join(',')}`
        }

        // Add period (pe) to iframeSrc
        if (itemFilters[FILTER_PE]?.length) {
            const peValues = itemFilters[FILTER_PE].map((x) => x.id).join(',')
            iframeSrc += `&period=${peValues}`
        }

        return iframeSrc
    }
}
```

jenniferarnesen commented:
@EdsonNhancale Could you also confirm whether the `Known Issue` in the PR description can be checked off?

KaiVandivier commented:
Perhaps this case is covered by the new ability to add custom app plugins to the Dashboard now?

EdsonNhancale committed:
fix: merged getIframeSrc.js file with master branch

jenniferarnesen commented:
> Perhaps this case is covered by the new ability to add custom app plugins to the Dashboard now?

This would be adding period filtering support for the older pre-existing dashboard widgets, which are different than the new plugin architecture. Since the older widgets will probably be around for some time, we should support it.

jenniferarnesen commented:
Superceded by https://github.com/dhis2/dashboard-app/pull/3255

Thank you for your contribution https://github.com/EdsonNhancale
